bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/RowResultAdapter.java

```diff
@@ -47,59 +46,43 @@
  * on duplicate labelled cells for its implementation. So this adapter will not deduplicate labelled
  * cells.
  *
+ * <p>This adapter will also return and check for scan marker rows, which will be an empty row with
+ * the scan marker row label.
+ *
  * <p>For internal use only - public for technical reasons.
  */
 @InternalApi("For internal usage only")
 public class RowResultAdapter implements RowAdapter<Result> {
 
   private static final byte[] EMPTY_VALUE = new byte[0];
+  private static final String SCAN_MARKER_ROW_LABEL = "bigtable-scan-marker-row";
 
   @Override
   public RowBuilder<Result> createRowBuilder() {
     return new RowResultBuilder();
   }
 
+  /**
+   * Checks if the result is a scan marker row. Returns true if the row's family, qualifier, and
+   * value are empty, and only has a scan marker row label.
+   */
   @Override
   public boolean isScanMarkerRow(Result result) {
-    Preconditions.checkState(result instanceof RowResult, "Should be an instance of RowResult");
-    return ((RowResult) result).isMarkerRow();
+    if (result.rawCells().length != 1 || !(result.rawCells()[0] instanceof RowCell)) {
+      return false;
+    }
+
+    RowCell cell = (RowCell) result.rawCells()[0];
+    return cell.getLabels().size() == 1
+        && cell.getLabels().get(0).equals(SCAN_MARKER_ROW_LABEL)
+        && cell.getValueArray().length == 0
+        && cell.getFamilyArray().length == 0
+        && cell.getQualifierArray().length == 0;
   }
 
   @Override
   public ByteString getKey(Result result) {
-    Preconditions.checkState(result instanceof RowResult, "Should be an instance of RowResult");
-    return ((RowResult) result).getKey();
-  }
-
-  @VisibleForTesting
-  static class RowResult extends Result {
-    private final ByteString rowKey;
-    private final boolean isMarkerRow;
-
-    static RowResult create(ByteString rowKey, List<Cell> cells) {
-      return new RowResult(rowKey, false, cells);
-    }
-
-    static RowResult createMarker(ByteString rowKey) {
-      return new RowResult(rowKey, true, ImmutableList.<Cell>of());
-    }
-
-    private RowResult(ByteString rowKey, boolean isMarkerRow, List<Cell> cells) {
-      this.rowKey = rowKey;
-      this.isMarkerRow = isMarkerRow;
-
-      // Only default ctor of Result is public, so instantiating cells through copyFrom() because
-      // value(), size(), isEmpty() rawCells() etc. directly usages Result's cells field.
-      this.copyFrom(Result.create(cells));
-    }
-
-    ByteString getKey() {
-      return rowKey;
-    }
-
-    boolean isMarkerRow() {
-      return isMarkerRow;
-    }
+    return ByteStringer.wrap(result.rawCells()[0].getRowArray());
```

Comment:
I think this makes an assumption that is not in line with the contract of a Cell. You are assuming that getRowOffset() == 0 and getRowLength() == getRowArray().length. The RowCell implementation aligns with this, but there is nothing in this method to ensure that implementation. I think it would be a lot safer to this instead:
```
Cell c = result.rawCells()[0];
ByteString.wrap(c.getRowArray(), c.getRowOffset(), c.getRowLength()
```